### ISSUE TITLE
S3 compatibility with AWS Cli and AIS bucket

### ISSUE THREAD

**Issue by Timshel** [NVIDIA/aistore#67]
Hi,

I started to test aistore recently, I had some issue with the quickstart so I just reused the [docker image](https://github.com/NVIDIA/aistore/blob/master/deploy/dev/k8s/Dockerfile-aisnode-ubuntu) from dev k8s templates with minor modification (updated to ubuntu 20.04 and cloned the `3.2.1` project version instead, cf [gist](https://gist.github.com/Timshel/9af876f7c00127b3807854dcf368947d)).

I then started a proxy and target node on two different node using host network, cluster initialization appear to be ok and `ais show cluster` return both nodes as healthy.

First test I wanted to do then is the S3 compatibility, so I created a bucket using `ais create bucket test` and pushed a file in it. I then tried to interact with using  `aws  --endpoint-url http://10.10.1.21:5353/s3 s3 ls s3://`., it returned the bucket list. But when I started to try to interact with the bucket I had multiple issues :

- on `aws --end... s3 ls s3://test/` : `Invalid timestamp "": String does not contain a date`:  using `tcpdump` I can see that the `LastModified` is empty
- on `aws --end... s3 cp s3://test/Chart-88.png  .`:  `fatal error: Could not connect to the endpoint URL: http://s3.ais.amazonaws.com/s3/test/Chart-88.png` (Not sure where the `s3.ais.amazonaws.com` come from it's not in my config and using `tcpdump` I can see that the proxy return a 307 to the `target` with the correct url).
- using `s3cmd` I was able to push a file even if the command line returned errors `MD5 Sums don't match!` (I downloaded it afterward and it has the same sha256 hash).

I tested using awscli `1.18.186` and `2.1.1` (I use some random access key/secret but disabled the auth).

Is this expected or do I have issues with my setup ?
Thank you and have a nice day :).

**Comment by VladimirMarkelov** [comment]
It is known issue. Basic operations like `ls` should work(I'll check why it complains about empty time - it must work), but more complex ones like `cp` won't. When I debugged `cp` I found that `aws` CLI contains hardcoded addresses `amazonaws.com`. `aws` adds it to your addresses passed in `--end` argument, and it was unavoidable. It made all `cp` calls to fail. When I asked at the library forum, I was told that the library used by `aws` was not intended to work with custom URLs, and if I would like to test my own S3 implementation I would have to use another utility like `s3cmd`.

As of  checksum, default checksum for AIS is `xxhash`, while `s3cmd` expects `md5`. Here is the trouble. You can try changing bucket's checksum algorithm: `ais set props <bucket-name> checksum.type=md5`:

```shell
$ ais create bucket ais://bck1
$ ais set props ais://bck1 checksum.type=md5
$ ais put go.sum ais://bck1/gsum
$ ais ls ais://bck1/ --props all
NAME     SIZE            CHECKSUM                                ATIME                   VERSION         CACHED  TARGET URL              STATUS  COPIES
gg       69.99KiB        a103a20a4e8a207fe7ba25eeb2634c96        02 Dec 20 09:54 PST     1               yes     http://127.0.0.1:8085   ok      1
$ md5sum go.sum
a103a20a4e8a207fe7ba25eeb2634c96  go.sum
```

Checksums equal.

**Comment by VladimirMarkelov** [comment]
Sorry, missed to mention that I checked everything on the latest master (a03a3165). We are about to release v3.3 - earliest this Fri. You could, therefore, get and check the master branch which at this point must be stable. 

As of `Atime`, I tried it a few times but failed to reproduce it on my machine. It would be great if you could collect the logs from your cluster and attach them to the issue. I'll look through them in case I missed anything and there is a hidden trouble.

PUT must work on the latest master as well. Here is the log of one of my attempts (and I'll explain how to fix "checksum mismatch" as well):

```
# I deployed a fresh cluster
$ ais create bucket ais://bck
"ais://bck" bucket created

$ s3cmd --host=https://localhost:8080/s3 put go.sum s3://bck/gsum-fail
upload: 'go.sum' -> 's3://bck/gsum-fail'  [1 of 1]
 71671 of 71671   100% in    0s    25.33 MB/s  done
 ... a few lines skipped ...
WARNING: MD5 Sums don't match!
WARNING: Too many failures. Giving up on 'go.sum'
ERROR: Upload of 'go.sum' failed too many times (Last reason: )

# let's change the bucket checksum
$ ais set props ais://bck checksum.type=md5
Bucket props successfully updated
"checksum.type" set to:"md5" (was:"xxhash")
$ s3cmd --host=https://localhost:8080/s3 put go.sum s3://bck/gsum-md5
upload: 'go.sum' -> 's3://bck/gsum-md5'  [1 of 1]
 71671 of 71671   100% in    0s    25.43 MB/s  done
upload: 'go.sum' -> 's3://bck/gsum-md5'  [1 of 1]
 71671 of 71671   100% in    0s    10.47 MB/s  done

# List the bucket with `ais` - note the difference in version numbers and checksums
$ ais ls ais://bck --props all
NAME             SIZE            CHECKSUM                                ATIME                   VERSION         CACHED  TARGET URL              STATUS  COPIES
gsum-fail        69.99KiB        45b8d688183150a6                        02 Dec 20 17:22 PST     6               yes     http://127.0.0.1:8087   ok      1
gsum-md5         69.99KiB        a103a20a4e8a207fe7ba25eeb2634c96        02 Dec 20 17:23 PST     1               yes     http://127.0.0.1:8088   ok      1

# Try to download both with `s3cmd`
$ s3cmd --host=https://localhost:8080/s3 get s3://bck/gsum-fail gosum.fail
download: 's3://bck/gsum-fail' -> 'gosum.2'  [1 of 1]
download: 's3://bck/gsum-fail' -> 'gosum.2'  [1 of 1]
 71671 of 71671   100% in    0s    61.03 MB/s  done
WARNING: MD5 signatures do not match: computed=a103a20a4e8a207fe7ba25eeb2634c96, received=
$ s3cmd --host=https://localhost:8080/s3 get s3://bck/gsum-md5 gosum.md
download: 's3://bck/gsum-md5' -> 'gosum.1'  [1 of 1]
download: 's3://bck/gsum-md5' -> 'gosum.1'  [1 of 1]
 71671 of 71671   100% in    0s    57.01 MB/s  done
```

**Comment by Timshel** [comment]
Thank you for the quick responses :).

For the `amazonaws.com` subject yes I supposed it was so; just want to mention I have no idea which kind of hacks they had to do but the [MinIO](https://github.com/minio/minio) project works with `awscli` (maybe some tricks could be ported ;).

I tested earlier the `put` but still had the same issue even after changing the `checksum.type` (I checked it was ok in the `props`). Running `tcpdump` I could see that the response was 200 with just the`Content-Length: 0` header. I believe the client then complained since it could not find the `ETag` header which I believe contain the hash of the part just uploaded.

I will try again tomorrow against the latest master.
